flink-streaming-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -17,157 +17,553 @@
+		}
 
-					// check if there are consumer threads still running
-					consumersStillRunning = false;
-					for (ShardConsumerThread<?> consumerThread : consumerThreads) {
-						consumersStillRunning = consumersStillRunning | consumerThread.isAlive();
-					}
-				} catch (InterruptedException e) {
-					// ignore
-				}
+		//  3. start consuming any shard state we already have in the subscribedShardState up to this point; the
+		//     subscribedShardState may already be seeded with values due to step 1., or explicitly added by the
+		//     consumer using a restored state checkpoint
+		for (int seededStateIndex = 0; seededStateIndex < subscribedShardsState.size(); seededStateIndex++) {
+			KinesisStreamShardState seededShardState = subscribedShardsState.get(seededStateIndex);
+
+			if (LOG.isInfoEnabled()) {
+				LOG.info("Subtask {} will start consuming seeded shard {} from sequence number {} with ShardConsumer {}" +
+					runtimeContext.getIndexOfThisSubtask(), seededShardState.getKinesisStreamShard().toString(),
+					seededShardState.getLastProcessedSequenceNum(), seededStateIndex);
 			}
 
-			// make sure any asynchronous error is noticed
-			Throwable error = this.error.get();
-			if (error != null) {
-				throw new Exception(error.getMessage(), error);
+			shardConsumersExecutor.submit(
+				new ShardConsumer<>(
+					this,
+					seededStateIndex,
+					subscribedShardsState.get(seededStateIndex).getKinesisStreamShard(),
+					subscribedShardsState.get(seededStateIndex).getLastProcessedSequenceNum()));
+		}
+
+		// ------------------------------------------------------------------------
+
+		// finally, start the infinite shard discovery and consumer launching loop;
+		// we will escape from this loop only when shutdownFetcher() or stopWithError() is called
+
+		final long discoveryIntervalMillis = Long.valueOf(
+			configProps.getProperty(
+				KinesisConfigConstants.CONFIG_SHARD_DISCOVERY_INTERVAL_MILLIS,
+				Long.toString(KinesisConfigConstants.DEFAULT_SHARD_DISCOVERY_INTERVAL_MILLIS)));
+
+		while (running) {
+			List<KinesisStreamShard> newShardsDueToResharding = discoverNewShardsToSubscribe();
+
+			for (KinesisStreamShard shard : newShardsDueToResharding) {
+				// since there may be delay in discovering a new shard, all new shards due to
+				// resharding should be read starting from the earliest record possible
+				KinesisStreamShardState newShardState =
+					new KinesisStreamShardState(shard, SentinelSequenceNumber.SENTINEL_EARLIEST_SEQUENCE_NUM.get());
+				int newStateIndex = registerNewSubscribedShardState(newShardState);
+
+				if (LOG.isInfoEnabled()) {
+					LOG.info("Subtask {} has discovered a new shard {} due to resharding, and will start consuming " +
+							"the shard from sequence number {} with ShardConsumer {}",
+						runtimeContext.getIndexOfThisSubtask(), newShardState.getKinesisStreamShard().toString(),
+						newShardState.getLastProcessedSequenceNum(), newStateIndex);
```

Comment:
I would add a LOG.debug statement everytime we call `discoverNewShardsToSubscribe()`.